## <font color=red>Литература</font>

1. Лекции «Алгоритмы: построение, анализ и реализация на языке программирования Си» - Ворожцов А.В., Винокуров Н.А.. Выложена в канале **#edu_materials**


 # <font color=blue>Алгоритмы сортировки, не использующие сравнение 2</font>


## <font color=green>Поразрядная сортировка (radix sort)</font>

Поразрядная сортировка - это развитие идей сортировки подсчетом и быстрой сортировки. Пусть ключи, по которым осуществляется сортировка - целые числа. Тогда можно разделить массив на корзины по значениям старших разрядов ключей. Затем каждую из получившихся корзин разделить уже по значениям второго по старшинству разряда и так далее. Подход очень похож на быструю сортировку, однако на практике почти всегда под корзины выделяется дополнительная память, что обеспечивает сортировке стабильность.

Поразрядная сортировка есть в двух вариантах: по более значимым цифрам (most significant digit или MSD) и по менее значимым цифрам (least significant digit или LSD). Сверху описан алгоритм MSD.

Алгоритм LSD устроен следующим образом.

1. Выполняется **устойчивая** (т.е. равные элементы не переставляются) сортировка подсчетом по младшему разряду.

2. Корзины соединяются (конкатенируются) в один список.

3. Повторяется пункт 1, но сортировка выполняется по следующему по старшинству разряду.

### Пример 1. LSD сортировка 20-ти трехзначных чисел

```
117 821 112 951 340 832 417 392 210 252 939 962 937 715 481 132 192 641 791 988  # исходный массив
340 210 821 951 481 641 791 112 832 392 252 962 132 192 715 117 417 937 988 939  # упорядочен младший разряд
210 112 715 117 417 821 832 132 937 939 340 641 951 252 962 481 988 791 392 192  # упорядочены 0-й и 1-й разряды
112 117 132 192 210 252 340 392 417 481 641 715 791 821 832 937 939 951 962 988
```

***
Время поразрядной сортировки оценивается в $O\left(w \cdot (n + d)\right)$, где $w$ - максимальная длина ключа.

| <font size=3>Сложность</font> | <font size=3>Лучший случай</font> |<font size=3>Средняя эффективность</font>|<font size=3>Худший случай</font>|  
| :---: | :---: | :---: | :---: |
| <font size=3>Время</font> | <font size=3> $O\left(w \cdot (n + d)\right)$ </font> | <font size=3> $O\left(w \cdot (n + d)\right)$ </font> | <font size=3> $O\left(w \cdot (n + d)\right)$ </font> |
| <font size=3>Память MSD</font> | <font size=3>$O(n + d \cdot w)$</font> | <font size=3>$O(n + d \cdot w)$</font> | <font size=3>$O(n + d \cdot w)$</font> |
| <font size=3>Память LSD</font> | <font size=3>$O(n + d)$</font> | <font size=3>$O(n + d)$</font> | <font size=3>$O(n + d)$</font> |

### Упражнение 4. LSD sorting

Реализуйте функцию `lsd_sort()` для поразрядной сортировки.

In [2]:
import random
L = list(range(150)) + list(range(100, 250))
random.shuffle(L)

In [3]:
# L - сортируемый список; i - номер разряда
def count_sort(L, i):
    
    # Словарь цифра в разряде -> элементы с одинаковым значением этого разряда
    count = {k: [] for k in range(10)}
    count['empty'] = []
        
    for elem in L:
        if elem // 10**i == 0:
            count['empty'].append(elem)
        else:
            digit = (elem // 10**i) % 10
            count[digit].append(elem)
        
    L[: len(count['empty'])] = count['empty']
    index = len(count['empty'])
    for j in range(10):
        L[index: index + len(count[j])] = count[j]
        index += len(count[j])

In [4]:
def radix_sort(L):
    
    # Максимальное число разрядов
    num_digits = len(str(max(L)))
    
    # Для каждого разряда с младшего к старшему
    for i in range(num_digits):
        
        # Для каждого разряда вызываем сортировку подсчетом
        count_sort(L, i)

In [5]:
L_sorted = sorted(L)

radix_sort(L)

print(L_sorted == L)

True


In [6]:
L

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 100,
 101,
 101,
 102,
 102,
 103,
 103,
 104,
 104,
 105,
 105,
 106,
 106,
 107,
 107,
 108,
 108,
 109,
 109,
 110,
 110,
 111,
 111,
 112,
 112,
 113,
 113,
 114,
 114,
 115,
 115,
 116,
 116,
 117,
 117,
 118,
 118,
 119,
 119,
 120,
 120,
 121,
 121,
 122,
 122,
 123,
 123,
 124,
 124,
 125,
 125,
 126,
 126,
 127,
 127,
 128,
 128,
 129,
 129,
 130,
 130,
 131,
 131,
 132,
 132,
 133,
 133,
 134,
 134,
 135,
 135,
 136,
 136,
 137,
 137,
 138,
 138,
 139,
 139,
 140,
 140,
 141,
 141,
 142,


### <font color=red>Замечание</font>

Этот способ теоретически может оказаться быстрее алгоритма быстрой сортировки, однако в практике такие случаи встречаются нечасто. Чтобы поразрядная сортировка была эффективной требуется, чтобы ключи были равномерно распределены по цифрам в каждом из своих разрядов. Например, требуется отсортировать массив целых чисел 110, 110, 111, 112, 115, 111, 119, 118, 119, 117, 120, 116, 116, 112, 113, 114, 113, 117. В данном случае ключи распределены по цифрам в нулевом разряде следующим образом:
```
'0': 3
'1': 2
'2': 2
'3': 2
'4': 1
'5': 1
'6': 2
'7': 2
'8': 1
'9': 2
```
В первом разряде используются только два значения из 10-ти возможных: `'1'` и `'2'`, причем последнее встречается лишь один раз, а во втором разряде используется только значение `'1'`. Получается, что при использовании поразрядной сортировки на двух разрядах алгоритм будет "простаивать".

# <font color=blue>Свойства и классификация алгоритмов сортировки</font>

## <font color=green>Устойчивость</font>

Алгоритмы сортировки дялятся на устойчивые и неустойчивые. Устойчивые алгоритмы сохраняют порядок элементов с одинаковыми ключами.

### Пример 2. Устойчивость

Пусть требуется упорядочить слова в списке `["Bob", "Ann", "Angie"]` по первой букве. Ключи слов `"Ann"` и `"Angie"` совпадают и равны `"A"`. Если сортировка устойчива, то гарантируется, что в упорядоченном списке `"Ann"` будет стоять перед `"Angie"`.

### Упражнение 5. Устойчивость сортировки выбором

Почему сортировка выбором в варианте, когда выполняется перестановка элементов, не является устойчивой? Приведите пример. Реализуйте устойчивый вариант сортировки выбором. Для этого нужно вставлять минимальный элемент перед неупорядоченной частью массива.

### Упражнение 6. Устойчивость сортировки слиянием

Какая неточность в реализации сортировки слиянием приведет к тому, что она станет неустойчивой?

### Упражнение 7. Устойчивость быстрой сортировки

Почему быстрая сортировка неустойчива? Приведите пример. Реализуйте устойчивый вариант быстрой сортировки. Для этого нужно закрепить за каждым элементом массива его индекс в исходном массиве и в случае, когда ключи равны, большим будет считаться элемент с большим индексом.

**Вариант быстрой сортировки, в котором ключ расширяется засчет индекса элемента в исходном массиве требует дополнительно $O(n)$ памяти.**

## <font color=green>Естественность поведения</font>

Алгоритм принято называть естественным, если он эффективно работает на упорядоченных и частично упорядоченных массивах. Алгоритмы сортировки пузырьком и вставками эффективны на частично упорядоченных массивах, в то время как поведение быстрой сортировки и сортировки слиянием не естественно.

## <font color=green>Внутренняя и внешняя сортировка</font>

Сортировка называется **внутренней**, если массив целиком помещается в оперативной памяти и возможен доступ к произвольному элементу массива. Сортировка называется **внешней**, если данные находятся на носителе большого объема с последовательным доступом к памяти. На этом занятии рассматриваются только алгоритмы внутренней сортировки.

### Упражнение 8. Рекламная акция

В сети магазинов "Н-Аудио" проводится рекламная акция. Каждый второй товар – бесплатно. Естественно, кассирам дано указание пробивать товары в таком порядке, чтобы магазин потерял как можно меньше денег.

По списку товаров определите максимальную сумму в чеке.

**Вход**: натуральное число товаров (N < 1000) и далее N натуральных чисел – цены товаров.

**Выход**: одно число – максимальная сумма чека.

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='border-collapse:collapse;border:none;mso-border-alt:solid windowtext .5pt;
 mso-yfti-tbllook:480;mso-padding-alt:0cm 5.4pt 0cm 5.4pt'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width=121 valign=top style='width:90.45pt;border:solid windowtext 1.0pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Вход</p>
  </td>
  <td width=132 valign=top style='width:99.25pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Выход</p>
  </td>
  <td width=385 valign=top style='width:288.85pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Комментарий</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td width=121 valign=top style='width:90.45pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>5 2 1 10 50 10</p>
  </td>
  <td width=132 valign=top style='width:99.25pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>70</p>
  </td>
  <td width=385 valign=top style='width:288.85pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Возможен такой порядок: 10 2 50 1 10</p>
  </td>
 </tr>
</table>


### Упражнение 9. Ровно K обменов

Даны два числа N и К. Вывести перестановку из N чисел (N чисел от 1 до N без повторений) такую чтобы сортировка пузырьком по возрастанию на соответствующем массиве сделала ровно K обменов. Если возможных ответов несколько – выведите любой.

**Вход**: натуральное число N (N < 100) и целое  неотрицательное К. Гарантируется, что для всех наборов тестовых данных решение существует.

**Выход**: ответ на задачу.

<table class=MsoTableGrid border=1 cellspacing=0 cellpadding=0
 style='border-collapse:collapse;border:none;mso-border-alt:solid windowtext .5pt;
 mso-yfti-tbllook:480;mso-padding-alt:0cm 5.4pt 0cm 5.4pt'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes'>
  <td width=177 valign=top style='width:133.0pt;border:solid windowtext 1.0pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Вход</p>
  </td>
  <td width=118 valign=top style='width:88.4pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Выход</p>
  </td>
  <td width=343 valign=top style='width:257.15pt;border:solid windowtext 1.0pt;
  border-left:none;mso-border-left-alt:solid windowtext .5pt;mso-border-alt:
  solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>Комментарий</p>
  </td>
 </tr>
 <tr style='mso-yfti-irow:1;mso-yfti-lastrow:yes'>
  <td width=177 valign=top style='width:133.0pt;border:solid windowtext 1.0pt;
  border-top:none;mso-border-top-alt:solid windowtext .5pt;mso-border-alt:solid windowtext .5pt;
  padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>5 3</p>
  </td>
  <td width=118 valign=top style='width:88.4pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>3 1 2 5 4</p>
  </td>
  <td width=343 valign=top style='width:257.15pt;border-top:none;border-left:
  none;border-bottom:solid windowtext 1.0pt;border-right:solid windowtext 1.0pt;
  mso-border-top-alt:solid windowtext .5pt;mso-border-left-alt:solid windowtext .5pt;
  mso-border-alt:solid windowtext .5pt;padding:0cm 5.4pt 0cm 5.4pt'>
  <p class=MsoNormal>В процессе сортировки массива 3 1 2 5 4 будут совершены
  обмены: 1<span style='font-family:Symbol;mso-ascii-font-family:"Times New Roman";
  mso-hansi-font-family:"Times New Roman";mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>«</span></span>3
  3<span style='font-family:Symbol;mso-ascii-font-family:"Times New Roman";
  mso-hansi-font-family:"Times New Roman";mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>«</span></span>2
  и 5<span style='font-family:Symbol;mso-ascii-font-family:"Times New Roman";
  mso-hansi-font-family:"Times New Roman";mso-char-type:symbol;mso-symbol-font-family:
  Symbol'><span style='mso-char-type:symbol;mso-symbol-font-family:Symbol'>«</span></span>4.
  Возможны и другие варианты ответа.</p>
  </td>
 </tr>
</table>


### Упражнение 10. Лента Мебиуса

Предположим, что имеется некоторый кусок ленты, разделенный на кадры. Кадры занумерованы с двух сторон. Полоска ленты склеена в лист Мебиуса. Необходимо составить алгоритм упорядочения этой последовательности, предположив, что соседние кадры можно переставлять, (естественно, в упорядоченной последовательности будет один "скачок" от минимального элемента к максимальному). Следует учесть, что при перестановке кадров переставляются числа с обеих сторон кадров. 

Числа на противоположных сторонах кадра могут не совпадать. Всегда ли возможно такое упорядочение?